messaging/src/main/java/org/axonframework/commandhandling/gateway/DefaultCommandGateway.java

```diff
@@ -141,6 +150,21 @@ public static Builder builder() {
         return result;
     }
 
+    private class CommandCallbackWrapper<C, R> implements CommandCallback<C, R> {
```

Comment:
I had a somewhat different approach in mind, so hear me out and let's discuss.

I assumed we'd make this a separate, perhaps `abstract` class.
The `FailureLoggingCallback` would `extend` the `CommandCallbackWrapper` implementation, making it "wrappable".

Furthermore, doing so would allow others to provide another "wrappable" `CommandCallabck`, that in turn could wrap another callback (etc).
So, what do you think?

Reply:
I'm afraid that would not work. As soon as you need to create a new instance for the callback to be wrapped, supplying a concrete class won't do. Unless we go back to the customizer/factory approach. 

Let me take you through it. I created this abstract class:

```java
public abstract class WrappedCommandCallback<C, R> implements CommandCallback<C, R> {

    private final CommandCallback<C, R> delegate;

    protected WrappedCommandCallback(CommandCallback<C, R> delegate) {
        this.delegate = delegate;
    }

    @Override
    public void onResult(CommandMessage<? extends C> commandMessage,
                         CommandResultMessage<? extends R> commandResultMessage) {
        this.handle(commandMessage, commandResultMessage);
        if (delegate != null) {
            delegate.onResult(commandMessage, commandResultMessage);
        }
    }

    protected abstract void handle(CommandMessage<? extends C> commandMessage,
                                   CommandResultMessage<? extends R> commandResultMessage);
}
```

And made FailureLoggingCallback use that:

```java
public class FailureLoggingCallback<C, R> extends WrappedCommandCallback<C, R> {
    private final Logger logger;

    public FailureLoggingCallback(CommandCallback<C, R> delegate) {
        this(LoggerFactory.getLogger(FailureLoggingCallback.class), delegate);
    }
    
    public FailureLoggingCallback(Logger logger, CommandCallback<C, R> delegate) {
        super(delegate);
        this.logger = logger;
    }

    @Override
    protected void handle(CommandMessage<? extends C> commandMessage,
                          CommandResultMessage<? extends R> commandResultMessage) {
        commandResultMessage.optionalExceptionResult()
                            .ifPresent(cause ->
                                               logger.warn("Command '{}' resulted in {}({})",
                                                           commandMessage.getCommandName(),
                                                           cause.getClass().getName(),
                                                           cause.getMessage()));
    }
}
```

Now, when we want to make the class customizable, we need to supply a delegate, which will only be created during sending of the command. So it's not possible:

![image](https://user-images.githubusercontent.com/3278878/162425798-14b5dc63-57ac-4775-84a3-a2e99ab8dad2.png)

Reply:
Something like this should work though:

```java
@FunctionalInterface
public interface CommandCallback<C, R> {

    /**
     * Invoked when command handling execution is completed.
     *
     * @param commandMessage       the {@link CommandMessage} that was dispatched
     * @param commandResultMessage the {@link CommandResultMessage} of the command handling execution
     */
    void onResult(CommandMessage<? extends C> commandMessage, CommandResultMessage<? extends R> commandResultMessage);

    default CommandCallback<C, R> wrap(CommandCallback<C, R> delegate) {
        if(delegate == null) {
            return this;
        }
        return new WrappedCommandCallback<>(this, delegate);
    }

    class WrappedCommandCallback<C, R> implements CommandCallback<C, R> {

        private final CommandCallback<C, R> original;
        private final CommandCallback<C, R> delegate;

        protected WrappedCommandCallback(CommandCallback<C, R> original, CommandCallback<C, R> delegate) {
            this.original = original;
            this.delegate = delegate;
        }

        @Override
        public void onResult(CommandMessage<? extends C> commandMessage,
                             CommandResultMessage<? extends R> commandResultMessage) {
            delegate.onResult(commandMessage, commandResultMessage);
            original.onResult(commandMessage, commandResultMessage);
        }
    }
}
```

Now the builder can be like this:
![image](https://user-images.githubusercontent.com/3278878/162426932-49950cf6-66e8-4e27-a0fb-7bd8c2e94b39.png)

And the command can be wrapped quite nicely:

```java

        FutureCallback<Object, Object> callback = new FutureCallback<>();
        send(command, callback.wrap(commandCallback));
```

Reply:
This approach will also allow people to wrap their own callbacks quite easily